### PR TITLE
Switch to Nix 2.0

### PR DESCRIPTION
**Pull request by edolstra**
Hopefully we can switch to Nix 2.0 for NixOS 18.03. Nix 2.0 should be released in the next few weeks. This PR makes 2.0 the default (by temporarily pointing `nix` to `nixUnstable`) and updates `nixos-install` to work with it. It also adds a command `nixos-enter` that factors out the functionality of `nixos-install --chroot`.

### CODE REVIEW

**Comment by dezgeg** [comment]
Do these installer changes also fix https://github.com/NixOS/nixpkgs/issues/27307 ?

File: nixos/modules/installer/tools/nixos-enter.sh
Comment (Mic92) [inline nixos/modules/installer/tools/nixos-enter.sh]: Bind mounting `/etc/resolv.conf` would be nice to have, if this tool is supposed to be a general-purpose tool.

File: pkgs/tools/package-management/nix/default.nix
Comment (dtzWill) [inline pkgs/tools/package-management/nix/default.nix]: Not to nag, but just checking: is this intended to use perl bindings from nixStable (nix 2)?

**Comment by edolstra** [comment]
@dtzWill Thanks, fixed.

**Comment by edolstra** [comment]
@dezgeg Yes, the entire build is done in the target filesystem.

**Comment by edolstra** [comment]
Jobset: https://hydra.nixos.org/jobset/nixos/nix-2.0

File: pkgs/tools/package-management/nix/default.nix
Comment (dezgeg) [inline pkgs/tools/package-management/nix/default.nix]: BTW, not strictly related to the 2.0 changes, but using the `//` operator here doesn't work well with `overrideAttrs` IIRC. That is, `(nix.overrideAttrs (attrs: whatever)).perl-bindings` doesn't exist.

Comment (kballard) [inline pkgs/tools/package-management/nix/default.nix]: I thought Nix 2.0 removed the dependency on Perl?

Comment (dtzWill) [reply pkgs/tools/package-management/nix/default.nix]: It does, the perl bindings are for using Nix *from* perl.  Previously nix utilities and such were themselves implemented in perl, which is no longer the case.

**Comment by copumpkin** [comment]
Also relevant for 17.09: https://github.com/NixOS/nixpkgs/pull/35468

File: nixos/modules/installer/tools/nixos-install.sh
Comment (copumpkin) [inline nixos/modules/installer/tools/nixos-install.sh]: Ooh, does this make the channel optional? I was just thinking of adding that feature to make-disk-image.nix, since it's quite annoying in many situations.

**Comment by Ericson2314** [comment]
Looks like the hydra job (and of borg) have passed it?!?!?!

**Comment by edolstra** [comment]
@vcunat Jobset is all green now so this should be good to merge.

File: nixos/modules/installer/tools/nixos-install.sh
Comment (edolstra) [reply nixos/modules/installer/tools/nixos-install.sh]: Well, it was already optional since we had the `--no-channel-copy` flag.

Comment (copumpkin) [reply nixos/modules/installer/tools/nixos-install.sh]: Oh I guess I just haven't exposed that in the disk image machinery. I'll probably take a look after this is merged.

File: nixos/modules/installer/tools/nixos-enter.sh
Comment (edolstra) [reply nixos/modules/installer/tools/nixos-enter.sh]: Yeah that would be nice.

**Review by shlevy — APPROVED** [review]
:ship:

**Comment by Ericson2314** [comment]
....did vcunat want to have a final say on this, or @edolstra did you just ping him as one of the release managers?

**Comment by edolstra** [comment]
Yeah, it should be up to the release managers.

**Comment by vcunat** [comment]
If you feel it's ready, I believe you can go along.  I haven't found time to really read the changes yet.

**Comment by copumpkin** [comment]
I haven't had a chance to test the image building changes yet, but don't let me hold you up. If something goes wrong, I'll fix it or talk to you 😄

**Comment by 7c6f434c** [comment]
Now that the builds are performed in the target store, should it be documented that to have `/tmp` for the build also in the target filesystem, the user should set `TMPDIR`? (Or do I misunderstand the situation?)

File: pkgs/tools/package-management/nix/default.nix
Comment (vcunat) [reply pkgs/tools/package-management/nix/default.nix]: Well, I see it's wrong, as `nixStable` itself has no perl stuff inside, but I still don't know surely if it should be `nixStable.perl-bindings` or `perl-bindings { nix = nix1; }`.

**Review by vcunat — APPROVED** [review]
The binding issue seems a nitpick, really.  It's only referenced by hydra and that seems unlikely to be used with nix1 anymore...

I can't see any other problem.

File: pkgs/tools/package-management/nix/default.nix
Comment (Ericson2314) [reply pkgs/tools/package-management/nix/default.nix]: Yeah that looks like a mistake. That isn't even Nix 2's perl bindings.

**Comment by vcunat** [comment]
^^ hehe, GitHub confusing people again.  I have reported such case to them once or twice during the past year or two, but apparently it still happens sometimes.

**Comment by vcunat** [comment]
Possibly a problem: #36268

**Comment by dezgeg** [comment]
I think these changes caused https://github.com/NixOS/nixpkgs/issues/38477. Previously, a `mount --rbind /sys` was done, but no longer, so I bet `efivarfs` is not getting mounted inside the `nixos-enter` chroot anymore.

**Pull request merged**
